File: src/csharp/Foundgine.Core/Semantic/Authorization/SemanticAuthorizationEvidence.cs
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;
using Foundgine.Core.Semantic.IR;

namespace Foundgine.Core.Semantic.Authorization;

/// <summary>
/// Immutable authorization evidence bound to the exact semantic contract that
/// was evaluated. Evidence from one contract cannot be replayed against another.
/// </summary>
public sealed record SemanticAuthorizationEvidence(
    string ContractFingerprint,
    string AuthorizationFingerprint,
    long? AuthorizationVersion = null,
    string? AuthorizationAuthorityFingerprint = null)
{
    public void EnsureMatches(SemanticContractSnapshot contract)
    {
        ArgumentNullException.ThrowIfNull(contract);
        if (!string.Equals(ContractFingerprint, contract.ContractFingerprint, StringComparison.Ordinal))
            throw new SemanticAuthorizationException(
                $"Authorization evidence is bound to semantic contract '{ContractFingerprint}', " +
                $"but the supplied contract is '{contract.ContractFingerprint}'.");
    }

    public static SemanticAuthorizationEvidence Create(
        SemanticContractSnapshot contract,
        SemanticOperation authorizedOperation)
    {
        ArgumentNullException.ThrowIfNull(contract);
        ArgumentNullException.ThrowIfNull(authorizedOperation);

        var canonical = JsonSerializer.Serialize(authorizedOperation);
        var payload = $"contract={contract.ContractFingerprint}|operation={canonical}";
        return new SemanticAuthorizationEvidence(
            contract.ContractFingerprint,
            Convert.ToHexString(SHA256.HashData(Encoding.UTF8.GetBytes(payload))));
    }
}

File: src/java/foundgine-core/src/main/java/com/foundgine/core/semantic/authorization/SemanticAuthorizationEvidence.java
package com.foundgine.core.semantic.authorization;

import com.foundgine.core.semantic.*;
import com.foundgine.core.semantic.ir.*;
import com.foundgine.core.semantic.ir.graph.*;
import java.nio.charset.StandardCharsets;
import java.security.*;
import java.util.*;

/**
 * Immutable authorization evidence bound to the exact semantic contract
 * evaluated.
 */
public record SemanticAuthorizationEvidence(String contractFingerprint, String authorizationFingerprint,
		Long authorizationVersion, String authorizationAuthorityFingerprint) {
	public SemanticAuthorizationEvidence(String c, String a) {
		this(c, a, null, null);
	}

	public void ensureMatches(SemanticContractSnapshot c) {
		Objects.requireNonNull(c);
		if (!contractFingerprint.equals(c.contractFingerprint()))
			throw new SemanticAuthorizationException("Authorization evidence is bound to semantic contract '"
					+ contractFingerprint + "', but the supplied contract is '" + c.contractFingerprint() + "'.");
	}

	public static SemanticAuthorizationEvidence create(SemanticContractSnapshot c, SemanticOperation op) {
		Objects.requireNonNull(c);
		Objects.requireNonNull(op);
		var payload = "contract=" + c.contractFingerprint() + "|operation="
				+ SemanticOperationGraph.create(op).fingerprint();
		return new SemanticAuthorizationEvidence(c.contractFingerprint(), sha256(payload));
	}

	private static String sha256(String s) {
		try {
			var d = MessageDigest.getInstance("SHA-256").digest(s.getBytes(StandardCharsets.UTF_8));
			var b = new StringBuilder();
			for (byte x : d)
				b.append(String.format("%02x", x));
			return b.toString();
		} catch (Exception e) {
			throw new IllegalStateException(e);
		}
	}
}
